src/redis_memory.rs

```diff
@@ -0,0 +1,242 @@
+ * All operations are - and should be - atomic.
+ */
+impl<Address: Send + Sync + Hash + Eq + Debug + Clone + ToRedisArgs, const WORD_LENGTH: usize>
+    MemoryADT for RedisMemory<Address, WORD_LENGTH>
+{
+    type Address = Address;
+    type Error = RedisMemoryError;
+    type Word = [u8; WORD_LENGTH];
+
+    /**
+     * Atomically reads the values at the given addresses.
+     */
+    async fn batch_read(
+        &self,
+        addresses: Vec<Address>,
+    ) -> Result<Vec<Option<Self::Word>>, Self::Error> {
+        let safe_connection = &mut *self.connection.lock().expect("Poisoned lock.");
+        let refs: Vec<&Address> = addresses.iter().collect(); // Redis MGET requires references to the values
+        let res: Vec<Option<Self::Word>> =
+            safe_connection.mget(&refs).map_err(|e| RedisMemoryError {
+                details: e.to_string(),
+            })?;
+        Ok(res)
+    }
+
+    /**
+     * Atomically writes the bindings if the guard is still valid.
+     * Returns the current value on the guard's address.
+     * If the result is equal to the guard's old value, the bindings get
+     * written.
+     *
+     * Args that are passed to the LUA script are, in order :
+     * 1. Guard address.
+     * 2. Guard value.
+     * 3. Vector length.
+     * 4+. Vector elements (address, word).
+     */
+    async fn guarded_write(
+        &self,
+        guard: (Self::Address, Option<Self::Word>),
+        bindings: Vec<(Self::Address, Self::Word)>,
+    ) -> Result<Option<Self::Word>, Self::Error> {
+        let safe_connection = &mut *self.connection.lock().expect("Poisoned lock.");
+        let (guard_address, guard_value) = guard;
+
+        const GUARDED_WRITE_LUA_SCRIPT: &str = r#"
+        local guard_address = ARGV[1]
+        local guard_value = ARGV[2]
+        local length = ARGV[3]
+        
+        local value = redis.call('GET',ARGV[1])
+
+        -- compare the value of the guard to the currently stored value
+        if((value==false) or (not(value == false) and (guard_value == value))) then
+            -- guard passed, loop over bindings and insert them
+            for i = 4,(length*2)+3,2
+            do
+                redis.call('SET', ARGV[i], ARGV[i+1])
+            end
+        end
+        return value
+        "#;
+
+        let script = redis::Script::new(GUARDED_WRITE_LUA_SCRIPT);
+        let mut script_invocation = script.prepare_invoke();
+
+        script_invocation.arg(guard_address);
+        if let Some(byte_array) = guard_value {
+            script_invocation.arg(&byte_array);
+        } else {
+            script_invocation.arg("false".to_string());
+        }
+        script_invocation.arg(bindings.len());
+        for (address, word) in bindings {
+            script_invocation.arg(address).arg(&word);
+        }
+
+        let result: Result<Option<Self::Word>, redis::RedisError> =
+            script_invocation.invoke(safe_connection);
+
+        result.map_err(Into::into)
```

Comment:
This should be a one-liner.
- do not explicit the type
- use `.into()` (may not even be needed)

Reply:
New function
```
        script_invocation.invoke(&mut safe_connection).map_err(|e| e.into())
```

Reply:
ou ```
        script_invocation.invoke(&mut safe_connection).map_err(Into::into)```